### ISSUE TITLE
PBA network support

### ISSUE THREAD

**Issue by davidepic91** [sedutil/sedutil#23]
By reading your comments I understand that the PBA is based on buildroot and syslinux. The first one is used to automates the process of building a complete Linux system while the second manages the boot process. For an academic project, I need to add network functionalities to the actual configuration.

I read that syslinux does not provide a network stack. Despite this, I tried to use sockets in GetPassPhrase.cpp by simply adding "#include <sys/socket.h>" but it does not work, in fact it returns "coud not create socket:function not implemented". I note that in "sedutil-master/images/scratch/syslinux-6.03/core/lwip/src/include/lwip" there are some useful headers (sockets.h, tcp.h, etc.etc). Is it possible to use them in some way ? Any suggestion ? 

If the answer is no, I read that there is another version of syslinux called "gPXE" which allows to perform a network booting. We can find it in "sedutil-master/images/scratch/syslinux-6.03/gpxe". 
My question: is it possible to import only its network "functionalities" in syslinux?

**Comment by cristim** [comment]
Thanks for this question, and good luck with your project!

This may be a question better asked to the syslinux developers. I'm not very familiar with it so I may be wrong, but I don't think the network stack needs to be added/configured in the syslinux bootloader, instead I think that should be done in the buildroot image, so you may need to change build process to include this in the buildroot image.

But keep in mind that we're in the process of removing buildroot entirely, see #21 for more details. If you're planning to do any changes to the buildroot image, those would relatively soon be obsoleted, so I would prefer if you could help us in the longer-term work we're doing on #21, if your project timeline allows this kind of detour.

**Comment by davidepic91** [comment]
Unfortunately I need to do my project in these days therefore I cannot wait. I searched my error and I found it corresponds to the fact that there are not Unix domain sockets on kernel.

**Comment by davidepic91** [comment]
Is there a way to include in a static way (or something similar) all the socket stuff when I compile GetPassPhrase.cpp ? Or to inject them as the files in the "Common" folder ?

**Comment by cristim** [comment]
You will need to recompile the kernel used in the image, the kernel is currently compiled intentionally without network support. 

You can have a look at the kernel config used inside the PBA image: https://github.com/sedutil/sedutil/blob/master/images/buildroot/PBA64/kernel.config#L673

Later edit: you will also need to add some network card drivers, adding support for your hardware.

**Comment by cristim** [comment]
This is really getting out of scope of the project, for security reasons I don't think it makes sense to provide network support in the PBA, closing this as wontfix.

**Comment by davidepic91** [comment]
I understand when you say "security reasons" but there are scenarios in which boot from a network is very usefull. Thank you fot the help, you gave me very useful advice ;) . I will try to continue on my own.

**Comment by davidepic91** [comment]
I modified the ".config" of the Linux 4.3 Kernel from the "menuconfig". I added Network Support (TCP/IP stack and sockets) and the Device Driver for my Network Card but It not works. In practice I can create a socket but when I perform a connect it returns "Network unreachable". I'm sure that it is not a problem of my server because I can interact with it from another pc of my client. Maybe I'm not configuring the kernel in the right way. Is it possible that I must add a DHCP service or something similar ?

**Comment by cristim** [comment]
Yes, it seems you need to run a DHCP client, or to configure the connection manually.

**Comment by davidepic91** [comment]
Ok perfect. Thank you ! I notice that some days ago, the original project has been updated with some modifications to the buildroot system. I think it is important to inspect them before continue with this #21 ;)

**Comment by davidepic91** [comment]
I have another question :) . I added in "GetPassPhrase.cpp" two headers:
_#include <openssl/evp.h>
#include <openssl/rand.h>_
Now, If try to compile with the "pba" option on the docker, it returns 
_"In file included from GetPassPhrase.cpp:31:0:
openssl-aes.h:7:25: fatal error: openssl/evp.h: No such file or directory
#include <openssl/evp.h>"_
I tried to install openssl on the docker by modifying the "Dockerfile" in images. It installs them in fact I have:
_"Step 2/8 : RUN dpkg --add-architecture i386 && apt update && apt install -y         bc         build-essential         cpio         dosfstools         g++-multilib         gdisk         git-core         libncurses5-dev         libncurses5-dev:i386 	libssl-dev	openssl        python         squashfs-tools         sudo         unzip         wget "_
but I obtain the same error. it seems that it cannot find them during the compilation. Any suggestion ?

**Comment by davidepic91** [comment]
Solved. It works but I have a problem when I tried to modify the makefile in order to add a static library. In "sedutil-master/LinuxPBA/nbproject/Makefile-Release_x86_64.mk"  I change the following:

_# Link Libraries and Options
LDLIBSOPTIONS= -lcurses -ltinfo **-lcrypto -L/home/Desktop/BBB/sedutil-master/LinuxPBA libcrypto.a**_

lcrypto is the library for the headers while libcrypto.a is the library that I tried to link statically. It seems that in this way the static linking does not work.
